src/TensorTwoD.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+
+#include "config.h"
+
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <numeric>
+
+#include "TensorTwoD.hpp"
+
+#include "geopm/Exception.hpp"
+#include "TensorOneD.hpp"
+
+
+namespace geopm
+{
+
+
+    TensorTwoD::TensorTwoD(std::size_t rows, std::size_t cols)
+    {
+        set_dim(rows, cols);
+    }
+
+    TensorTwoD::TensorTwoD(const TensorTwoD &other)
+    {
+        m_mat = other.m_mat; 
+    }
+
+    TensorTwoD::TensorTwoD(json11::Json input)
+    {
+
+        if (!input.is_array()){
+            throw geopm::Exception("Neural network weights is non-array-type.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (input.array_items().size() == 0) {
+            throw geopm::Exception("Empty array is invalid for neural network weights.\n",
+                                   GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        set_dim(input.array_items().size(), input[0].array_items().size());
+        for (std::size_t idx = 0; idx < m_mat.size(); idx++) {
+            if (input[idx].array_items().size() != m_mat[idx].get_dim()) {
+                throw geopm::Exception("Attempt to load non-rectangular matrix.",
+                                       GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_mat[idx] = TensorOneD(input[idx]);
+        }
+    }
+
+    void TensorTwoD::set_dim(std::size_t rows, std::size_t cols)
+    {
+        //TODO: Should we be checking for negative values as well?
+        //And zeros?
+        if (rows == 0 && cols > 0) {
```

Comment:
What about: ``rows > 0 && cols == 0``?

I see TensorOneD accepts 0 as its size, however, I'm not seeing what you would be able to do with a TensorTwoD with those dimensions as it wouldn't be able to hold any values.